File: cplusplus/Core/DenseMatrix.cpp
/* DenseMatrix.cpp */
#include "DenseMatrix.h"
#include <stdexcept>
#include <cmath> //per std::abs in rank
#include <ctime> //time in setRandX
#include <cstdlib> //rand in setRandX


//Costruttore di default
DenseMatrix::DenseMatrix(unsigned int r, unsigned int c): MatrixBase(r, c), m(r, vector<double>(c)) {}

//Costruttore a 3 param
// crea una matrice con dim rows x cols
// imposta tutti gli elem a init
DenseMatrix::DenseMatrix(unsigned int r, unsigned int c, double init): MatrixBase(r, c), m(r, vector<double>(c, init)) {}

double& DenseMatrix::at(unsigned int i, unsigned int j) {
	return m.at(i).at(j);
}

const double& DenseMatrix::at(unsigned int i, unsigned int j) const {
	return m.at(i).at(j);
}

void DenseMatrix::swapRows(unsigned int row1, unsigned int row2, unsigned int fromCol) {
	for(unsigned int j=fromCol; j<numCols(); ++j) {
		double temp = at(row1, j);
		at(row1, j) = at(row2, j);
		at(row2, j) = temp;
	}
}

void DenseMatrix::swapCols(unsigned int col1, unsigned int col2, unsigned int fromRow) {
	for(unsigned int i=fromRow; i<numRows(); ++i) {
		double temp = at(i, col1);
		at(i, col1) = at(i, col2);
		at(i, col2) = temp;
	}
}

void DenseMatrix::print(ostream& os, string separator) const {
	for(unsigned int i=0; i<numRows(); ++i) {
		for(unsigned int j=0; j<numCols(); ++j)
			os << at(i,j) << separator;
		os << std::endl;
	}
}

void DenseMatrix::load(istream& is) {
	//controlla che sullo stream di input:
	// ci siano almeno i caratteri necessari per riempire la matrice (prima di eof)
	//	valori aggiuntivi (validi o meno) verranno ignorati
	// che tutti i caratteri necessari siano validi (double)
	//altrimenti:
	// eccezione e il contenuto della matrice è inconsistente (garbage)
	
	unsigned int i=0, j=0;
	bool stop = false;
	
	for( ; i<numRows() && !stop; ++i) {
		j=0;
		for( ; j<numCols() && !stop; ++j) {
			is >> at(i,j);
			if(is. eof() || is.fail()) //ho finito i dati o tra essi c'era un non-double
				stop = true;
		}
	}
	
	bool sufficienti = (i*j == numCols()*numRows());
	
	if( is.good() || (is.eof() && sufficienti) ) //tutto ok
        return;
    else
        if( is.eof() ) //&& !sufficienti
            throw std::invalid_argument("valori non sufficienti");
        else
            if( is.fail() ) //tra i dati che mi servono c'è un non-double
                throw std::invalid_argument("valori di tipo errato");
            else //is.bad()
                throw std::invalid_argument("errore imprevisto nelle fasi di input (bad stream), riavviare.\n Se il problema persiste contattare lo sviluppatore");
	

//	approccio naive che accetta tutto senza tanti problemi
//	for(unsigned int i=0; i<numRows(); ++i)
//		for(unsigned int j=0; j<numCols(); ++j)
//			is >> at(i,j);
	
}

//calcola il rango tramite il metodo di eliminazione di Gauss con pivoting parziale
unsigned int DenseMatrix::rank() const {
    unsigned int rig=numRows(), col=numCols(), rk=0;
	DenseMatrix* aux = clone(); //matrice ausiliaria d'appoggio
	bool finito = false;
	
	for(unsigned int d=0; d<col && d<rig && !finito; ++d) {
		
		//cerco il massimo pivot sotto di me
		unsigned int pivot_row = d;
		double cur = std::abs(aux->at(d,d));
		for(unsigned int k=d+1; k<rig; ++k)
			if(std::abs(aux->at(k,d)) > cur) {
				pivot_row = k;
				cur = aux->at(k,d);
			}

		 //se ne ho trovato uno maggiore, scambio la riga attuale con quella con il pivot massimo
		if(pivot_row != d)
			aux->swapRows(d, pivot_row, d); //solo gli elem da 'd' in poi, gli altri sono zeri
		
		if(std::abs(aux->at(d,d)) < 0.00000000000001) { //questa colonna ha solo zeri da qua in giù
			if(d+1<col) { //ho ancora colonne da controllare
				aux->swapCols(d, col-1, d); //col-1 è l'ultima, solo gli elem da 'd' in giù, gli altri non servono
				col--; //la colonna che sposto è inutile, quindi la escludo
				d--; //rifaccio questo giro d'iteraz, lo swap ha messo un nuovo elem in A[d][d]
			}
			else { //ho finito le colonne, quindi ho finito
				delete aux; //pulisco lo heap
				//return rk; //rk a questo punto sarà il num di col linear indip, cioè il rango
				finito=true; //se lascio qui dentro il return il compil mi da warning perchè non lo trova
			}
		}
		else { //aux->at(d,d)!=0
			//devo azzerare gli elem sotto di me, cioè:
			//aux->at(d+1,d), aux->at(d+2,d), ..., aux->at(i,d) con: d<i<rig
			for(unsigned int i=d+1; i<rig; ++i) {
				if(aux->at(i,d) != 0) { //va annullato
					//per annullarlo:
					//sostituisco l'intera riga in cui risiede l'elem da annullare (la i-esima)
					//con la differenza tra essa (la i-esima) e quella del 'pivot' (la d-esima)
					//moltiplicata per un coefficiente (coef)
					//coef è il rapporto tra l'elem da annullare (aux->at(i,d)) e il pivot (aux->at(d,d))
					double coef = aux->at(i,d) / aux->at(d,d);
					for(unsigned int j=d; j<col; ++j)
						aux->at(i,j) = aux->at(i,j) - (aux->at(d,j)*coef);
				}
				//else //l'elem è già zero, passo al prossimo
			}
			rk++; //questa, in cui A[d][d]!=0, è una colonna 'utile' che contribuisce al rango
		}
	}
	return rk;
}

void DenseMatrix::fill(double val) {
	for(unsigned int i=0; i<numRows(); ++i)
		for(unsigned int j=0; j<numCols(); ++j)
			at(i,j) = val;
}

void DenseMatrix::setRandInt(int from, int to) { //rand in [from,to)
	std::srand(std::time(0)); //seed
	for(unsigned int i=0; i<numRows(); ++i)
		for(unsigned int j=0; j<numCols(); ++j)
			at(i,j) = (std::rand() % (to-from) + from);
}

void DenseMatrix::setRandReal(double from, double to) { //rand in from[from,to]
	std::srand(std::time(0)); //seed
	for(unsigned int i=0; i<numRows(); ++i)
		for(unsigned int j=0; j<numCols(); ++j)
			at(i,j) = from + static_cast<double>(std::rand()) / (static_cast<double>(RAND_MAX/(to-from)));
		
}

File: java/matrixkalq/DenseMatrix.java
/* DenseMatrix.java */

package matrixkalq;

//import java.lang.Math;
//import java.lang.IllegalArgumentException;
//import java.lang.NumberFormatException;
import java.util.Vector;
import java.util.concurrent.ThreadLocalRandom;
import java.io.PrintStream;


//Abstract Base Class
public abstract class DenseMatrix extends MatrixBase {
	
	//Campi dati:
		private Vector<Vector<Double>> m;
		
	//Costruttori:
		public DenseMatrix(int r, int c, double init) {
			super(r,c);
			m = new Vector<Vector<Double>>(r);
			m.setSize(r);
			
			for(int i=0; i<r; ++i) {
				m.set(i, new Vector<Double>(c));
				m.get(i).setSize(c);
				for(int j=0; j<c; ++j)
					m.get(i).set(j,init);
			}
		}
		
		public DenseMatrix(int r, int c) {
			super(r,c);
			m = new Vector<Vector<Double>>(r);
			m.setSize(r);
			
			for(int i=0; i<r; ++i) {
				m.set(i, new Vector<Double>(c));
				m.get(i).setSize(c);
			}
		}
		
		public DenseMatrix() { this(0,0); }
		
	//Metodi astratti:
		public abstract DenseMatrix sum(MatrixBase m);
		public abstract DenseMatrix dif(MatrixBase m);
		public abstract DenseMatrix sProd(double s);
		public abstract DenseMatrix trans();
		
	//Override:
		public DenseMatrix clone() {
		
			DenseMatrix aux = (DenseMatrix)super.clone();
			
			int r=numRows(), c=numCols();
			Vector<Vector<Double>> m_aux = new Vector<Vector<Double>>(r);
			m_aux.setSize(r);
			for(int i=0; i<r; ++i) {
				m_aux.set(i, new Vector<Double>(c));
				m_aux.get(i).setSize(c);
				for(int j=0; j<c; ++j)
					m_aux.get(i).set(j, this.m.get(i).get(j));
			}
			aux.m = m_aux;
			
			return aux;
		}
		
		public String toString() {
			String aux = "";
			for(int i=0; i<numRows(); ++i) {
				for(int j=0; j<numCols(); ++j) {
						if(m.get(i).get(j) == null)
							aux = aux + "null" + "\t";
						else
							aux = aux + getElem(i,j) + "\t";
				}
				aux = aux + "\n";
			}
			return aux;
		}
		
	//Metodi 'concretizzati':
		public double getElem(int i, int j) { return m.get(i).get(j); }
		
		public void setElem(int i, int j, double val) { m.get(i).set(j,val); }
		
		public void swapRows(int row1, int row2, int fromCol) {
			for(int j=fromCol; j<numCols(); ++j) {
				double temp = getElem(row1,j);
				setElem(row1,j, getElem(row2,j));
				setElem(row2,j, temp);
			}
		}
		
		public void swapCols(int col1, int col2, int fromRow) {
			for(int i=fromRow; i<numRows(); ++i) {
				double temp = getElem(i,col1);
				setElem(i,col1, getElem(i,col2));
				setElem(i,col2, temp);
			}
		}
		
		public void print(PrintStream ps, String separ) {
			String aux = "";
			for(int i=0; i<numRows(); ++i) {
				for(int j=0; j<numCols(); ++j) {
						if(m.get(i).get(j) == null)
							aux = aux + "null" + separ;
						else
							aux = aux + getElem(i,j) + separ;
				}
				aux = aux + "\n";
			}
			ps.println(aux);
		}
		
		public void load(String s) {
			//controlla che sullo stream di input:
			// ci siano almeno i caratteri necessari per riempire la matrice
			//	valori aggiuntivi (validi o meno) verranno ignorati
			// che tutti i caratteri necessari siano validi (double)
			//altrimenti:
			// eccezione e il contenuto della matrice è inconsistente (garbage)
			
			if(s == null || s.isEmpty())
				s=" ";
			
			String[] temp = (s.replaceAll("\\s+"," ")).split(" ");
			
			if(temp.length < numRows()*numCols())
				throw new IllegalArgumentException("valori non sufficienti");
			else //ne ho abbastanza, verifico siano double e li carico
				for(int i=0; i<numRows(); ++i)
					for(int j=0; j<numCols(); ++j) {
						try { setElem(i,j, Double.parseDouble(temp[i*numCols()+j])); }
						catch(NumberFormatException e) { throw new NumberFormatException("valori di tipo errato"); }
					}
		}
		
		//Calcola il rango tramite il metodo di eliminazione di Gauss con pivoting parziale
		public int rank() {
			int rig=numRows(), col=numCols(), rk=0;
			DenseMatrix aux = clone(); //matrice ausiliaria d'appoggio
			boolean finito = false;
			
			for(int d=0; d<col && d<rig && !finito; ++d) {
				
				//cerco il massimo pivot sotto di me
				int pivot_row = d;
				double cur = Math.abs(aux.getElem(d,d));
				for(int k=d+1; k<rig; ++k)
					if(Math.abs(aux.getElem(k,d)) > cur) {
						pivot_row = k;
						cur = aux.getElem(k,d);
					}
				
				//se ne ho trovato uno maggiore, scambio la riga attuale con quella con il pivot massimo
				if(pivot_row != d)
					aux.swapRows(d, pivot_row, d); //solo gli elem da 'd' in poi, gli altri sono zeri
				
				if(Math.abs(aux.getElem(d,d)) < 0.00000000000001) { //questa colonna ha solo zeri da qua in giù
					if(d+1<col) { //ho ancora colonne da controllare
						aux.swapCols(d, col-1, d); //col-1 è l'ultima, solo gli elem da 'd' in giù, gli altri non servono
						col--; //la colonna che sposto è inutile, quindi la escludo
						d--; //rifaccio questo giro d'iteraz, lo swap ha messo un nuovo elem in A[d][d]
					}
					else { //ho finito le colonne, quindi ho finito
						finito=true; //rk a questo punto sarà il num di col linear indip, cioè il rango
					}
				}
				else { //aux.getElem(d,d)!=0
					//devo azzerare gli elem sotto di me, cioè:
					//aux.getElem(d+1,d), aux.getElem(d+2,d), ..., aux.getElem(i,d) con: d<i<rig
					for(int i=d+1; i<rig; ++i) {
						if(aux.getElem(i,d) != 0) { //va annullato
							//per annullarlo:
							//sostituisco l'intera riga in cui risiede l'elem da annullare (la i-esima)
							//con la differenza tra essa (la i-esima) e quella del 'pivot' (la d-esima)
							//moltiplicata per un coefficiente (coef)
							//coef è il rapporto tra l'elem da annullare (aux.getElem(i,d)) e il pivot (aux.getElem(d,d))
							double coef = aux.getElem(i,d) / aux.getElem(d,d);
							for(int j=d; j<col; ++j)
								aux.setElem(i, j, (aux.getElem(i,j) - (aux.getElem(d,j)*coef)));
						}
						//else //l'elem è già zero, passo al prossimo
					}
					rk++; //questa, in cui A[d][d]!=0, è una colonna 'utile' che contribuisce al rango
				}
			}
			return rk;
		}
		
	//Metodi propri:
		public void fill(double val) {
			for(int i=0; i<numRows(); ++i)
				for(int j=0; j<numCols(); ++j)
					setElem(i,j,val);
		}
		
		public void setRandInt(int from, int to) { //rand in [from,to)
			for(int i=0; i<numRows(); ++i)
				for(int j=0; j<numCols(); ++j)
					setElem(i, j, ThreadLocalRandom.current().nextInt(from, to));
		}
		
		public void setRandReal(double from, double to) { //rand in from[from,to)
			for(int i=0; i<numRows(); ++i)
				for(int j=0; j<numCols(); ++j)
					setElem(i, j, ThreadLocalRandom.current().nextDouble(from, to));
		}
		
	//Metodi ereditati:
		//	public int numRows();
		//	public int numCols();
		//	public boolean equals(Object obj);
		
	//Eccezioni:
		//	ArrayIndexOutOfBoundsException: getElem(..), setElem(..), swapRows(..), swapCols(..)
		//	NumberFormatException: load(..)
		//	IllegalArgumentException: load(..), setRandInt(..), setRandReal(..)
		//	RuntimeException: clone()
		
} //fine DenseMatrix
